File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;
int Max = 0;

bool check(int level,vector<int> diffs, vector<int> times, long long limit){
    long long sum = 0;
    int prev = 0;
    
    for(int i =0; i<diffs.size(); i++){
        if(diffs[i] <= level) sum += times[i];
        else{
            int X = diffs[i] - level;
            sum += (times[i] + prev) * X;
            sum += times[i];
        }
        if(sum > limit) return false;
        prev = times[i];
    }
    return true;
    
}

int solution(vector<int> diffs, vector<int> times, long long limit) {
    int answer = 0;
    
    for(auto c: diffs) Max = max(Max,c);
    
    int start = 1, end = Max;
    
    while(start <= end){
        int mid = (start + end)/2;
        bool clear = check(mid,diffs,times, limit);
        if(clear){
            answer = mid;
            end = mid - 1;
        }
        else start = mid + 1; 
    }

    
    return answer;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
class Solution {
    static int[] diff, time;
    static long l;
    public int solution(int[] diffs, int[] times, long limit) {
        int answer = 0;
        diff = diffs;
        time = times;
        l = limit;
        int start = 1;
        int end = 0;
        for(int diff : diffs) end = Math.max(end, diff);
        
        while(start < end){
            int mid = (start + end) / 2;
            if(check(mid)){
                end = mid;
            }
            else start = mid + 1;
        }
        answer = start;
        
        return answer;
    }
    static boolean check(int level){
        
        long t = 0;
        for(int i = 0; i < diff.length; i++){
            if(level >= diff[i]){
                t += time[i];
            }
            else{
                int re = diff[i] - level;
                t += (time[i] + time[i - 1]) * re + time[i];
            }
        }
        return (t <= l);
    }
}